File: python/백준/그래프/BFS/9205 맥주 마시면서 걸어가기.py
from collections import deque

T = int(input())

for _ in range(T):
    n = int(input())
    
    home = tuple(map(int, input().split()))
    conv = [tuple(map(int, input().split())) for _ in range(n)]
    festival = tuple(map(int, input().split()))
    
    queue = deque([home])
    visited = [False] * n
    
    is_happy = False
    
    while queue:
        x, y = queue.popleft()
        
        if abs(x - festival[0]) + abs(y - festival[1]) <= 1000:
            print("happy")
            is_happy = True
            break
        
        for i in range(n):
            if not visited[i]:
                nx, ny = conv[i]
                
                if abs(x - nx) + abs(y - ny) <= 1000:
                    visited[i] = True
                    queue.append((nx, ny))
    
    if not is_happy:
        print("sad")

File: cpp/백준/그래프/BFS/9205 맥주 마시면서 걸어가기.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <cstdlib>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;
    while (T--) {
        int n;
        cin >> n;

        pair<int, int> home;
        cin >> home.first >> home.second;

        vector<pair<int, int>> conv(n);
        for (int i = 0; i < n; ++i) cin >> conv[i].first >> conv[i].second;

        pair<int, int> festival;
        cin >> festival.first >> festival.second;

        queue<pair<int, int>> q;
        q.push(home);
        vector<bool> visited(n, false);
        bool is_happy = false;

        while (!q.empty()) {
            auto [x, y] = q.front();
            q.pop();

            if (abs(x - festival.first) + abs(y - festival.second) <= 1000) {
                is_happy = true;
                break;
            }

            for (int i = 0; i < n; ++i) {
                if (!visited[i]) {
                    int nx = conv[i].first;
                    int ny = conv[i].second;
                    if (abs(x - nx) + abs(y - ny) <= 1000) {
                        visited[i] = true;
                        q.push({nx, ny});
                    }
                }
            }
        }

        cout << (is_happy ? "happy" : "sad") << "\n";
    }
    return 0;
}
